File: packages/vm/wasmlib/go/wasmclient/clientfunc.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmclient

import "github.com/iotaledger/hive.go/crypto/ed25519"

type ClientFunc struct {
	svc     *Service
	keyPair *ed25519.KeyPair
	xfer    *Transfer
}

// Post sends a request to the smart contract service
// You can wait for the request to complete by using the returned Request
// as parameter to Service.WaitRequest()
func (f *ClientFunc) Post(hFuncName uint32, args *Arguments) Request {
	keyPair := f.keyPair
	if keyPair == nil {
		keyPair = f.svc.keyPair
	}
	return f.svc.PostRequest(hFuncName, args, f.xfer, keyPair)
}

// Sign optionally overrides the default keypair from the service
func (f *ClientFunc) Sign(keyPair *ed25519.KeyPair) {
	f.keyPair = keyPair
}

// Transfer optionally indicates which tokens to transfer as part of the request
// The tokens are presumed to be available in the signing account on the chain
func (f *ClientFunc) Transfer(xfer *Transfer) {
	f.xfer = xfer
}

File: packages/vm/wasmlib/ts/wasmclient/clientfunc.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmclient from "./index"
import {IKeyPair} from "./crypto";

export class ClientFunc {
    protected svc: wasmclient.Service;
    private keyPair: IKeyPair;
    private xfer: wasmclient.Transfer = new wasmclient.Transfer();

    constructor(svc: wasmclient.Service) {
        this.svc = svc;
    }

    // Sends a request to the smart contract service
    // You can wait for the request to complete by using the returned RequestID
    // as parameter to Service.waitRequest()
    public async post(hFuncName: wasmclient.Hname, args: wasmclient.Arguments | null): Promise<wasmclient.RequestID> {
        if (args == null) {
            args = new wasmclient.Arguments();
        }
        if (this.keyPair == null) {
            this.keyPair = this.svc.keyPair;
        }
        return await this.svc.postRequest(hFuncName, args, this.xfer, this.keyPair);
    }

    // Optionally overrides the default keypair from the service
    public sign(keyPair: IKeyPair): void {
        this.keyPair = keyPair;
    }

    // Optionally indicates which tokens to transfer as part of the request
    // The tokens are presumed to be available in the signing account on the chain
   public transfer(xfer: wasmclient.Transfer): void {
        this.xfer = xfer;
    }
}
